plan/gen_physical_plans.go

```diff
@@ -277,28 +286,98 @@ func (p *LogicalJoin) getIndexJoinByOuterIdx(prop *requiredProp, outerIdx int) [
-		matchedOffsets := joinKeysMatchIndex(innerJoinKeys, indexInfo)
-		if matchedOffsets == nil {
+		idxCols, colLengths := expression.IndexInfo2Cols(p.children[1-outerIdx].Schema().Columns, indexInfo)
+		if len(idxCols) == 0 {
+			continue
+		}
+		offsetsMap := joinKeysMatchIndex(innerJoinKeys, indexInfo)
+		if offsetsMap == nil {
 			continue
 		}
-		usedIndexInfo = indexInfo
-		newOuterJoinKeys := make([]*expression.Column, len(outerJoinKeys))
-		newInnerJoinKeys := make([]*expression.Column, len(innerJoinKeys))
-		for i, offset := range matchedOffsets {
-			newOuterJoinKeys[i] = outerJoinKeys[offset]
-			newInnerJoinKeys[i] = innerJoinKeys[offset]
+		maxIndexColIdx := 0
+		for i := len(indexInfo.Columns) - 1; i >= 0; i-- {
+			if offsetsMap[i] != -1 {
+				maxIndexColIdx = i
+				break
+			}
+		}
+		// After predicate push down, the one side conditions of join must be the conditions that cannot be pushed down so cannot
+		// to calculate range either.
+		// TODO: There may be a selection that block the index join.
+		conds := make([]expression.Expression, 0, len(offsetsMap)+len(x.pushedDownConds))
+		// Construct a fake equal expression for calculate the range.
+		for _, key := range innerJoinKeys {
+			fakeConstant := expression.One.Clone().(*expression.Constant)
+			fakeConstant.RetType = types.NewFieldType(key.RetType.Tp)
+			eqFunc := expression.NewFunctionInternal(p.ctx, ast.EQ, types.NewFieldType(mysql.TypeTiny), key, fakeConstant)
+			conds = append(conds, eqFunc)
+		}
+		lastExprHashCode := conds[offsetsMap[maxIndexColIdx]].HashCode()
+		conds = append(conds, x.pushedDownConds...)
+		// After constant propagation, there won'be cases that t1.a=t2.a and t2.a=1 occurs in the same time.
+		// So we can guarantee that once count of the columns in the ranges is no lower than the max idx
+		// that the inner keys matched in the index, then the equal conditions of join are all used.
+		accesses, remained := ranger.DetachIndexConditions(conds, idxCols, colLengths)
+		ranges, err := ranger.BuildRange(p.ctx.GetSessionVars().StmtCtx, accesses, ranger.IndexRangeType, idxCols, colLengths)
+		if err != nil {
+			terror.Log(errors.Trace(err))
+			continue
+		}
+		idxRanges := ranger.Ranges2IndexRanges(ranges)
+		// We should guarantee that all the index column in the join's equal condition is used.
+		// Only check the last one is ok.
+		valid := false
+		for _, expr := range accesses {
+			if bytes.Equal(expr.HashCode(), lastExprHashCode) {
+				valid = true
+				break
+			}
+		}
+		if !valid {
+			continue
+		}
+		// This compare way is a little easy, we can use the average unit size of one index in the future to determine which index to choose.
+		// There may be the cases like `a = 1 and b > 2 and b < 1`. So idxRanges can be nil though the conditions are valid.
+		if len(idxRanges) == 0 || len(idxRanges[0].LowVal) > maxRangeLen {
+			bestIndexInfo = indexInfo
+			if len(idxRanges) == 0 {
+				maxRangeLen = math.MaxInt32
```

Comment:
When will `len(idxRanges)` be 0?

Reply:
It can be. But we can ignore this case. See the new comment.